File: tools/shots.py
#!/usr/bin/env python3
"""Лист кадров сайта через headless Chromium (Playwright).
python3 shots.py <url> [--out DIR] [--w 1440] [--h 900] [--step 900] [--max 14] [--full]
Кадры пишутся в DIR (по умолчанию ./shots/<домен>/), имена 01.png, 02.png…
--full — дополнительно один кадр всей страницы целиком.
"""
import sys, os, re, time
from playwright.sync_api import sync_playwright

def main():
    a = sys.argv[1:]
    if not a: print(__doc__); return
    url = a[0]
    def opt(name, default):
        return a[a.index(name)+1] if name in a else default
    W, H = int(opt('--w', 1440)), int(opt('--h', 900))
    step, mx = int(opt('--step', 900)), int(opt('--max', 14))
    dom = re.sub(r'[^a-z0-9.-]', '_', re.sub(r'^https?://', '', url).split('/')[0])
    out = opt('--out', os.path.join('shots', dom)); os.makedirs(out, exist_ok=True)
    with sync_playwright() as p:
        b = p.chromium.launch()
        pg = b.new_page(viewport={'width': W, 'height': H}, device_scale_factor=2,
                        user_agent='Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 '
                                   '(KHTML, like Gecko) Chrome/124 Safari/537.36')
        pg.goto(url, wait_until='networkidle', timeout=60000)
        pg.wait_for_timeout(2500)
        total = pg.evaluate("document.body.scrollHeight")
        n = min(mx, max(1, -(-total // step)))
        files = []
        for i in range(n):
            pg.evaluate(f"window.scrollTo(0,{i*step})")
            pg.wait_for_timeout(1400)          # reveal-анимации
            f = os.path.join(out, f"{i+1:02d}.png"); pg.screenshot(path=f); files.append(f)
        if '--full' in a:
            f = os.path.join(out, "full.png"); pg.screenshot(path=f, full_page=True); files.append(f)
        print(f"высота {total}px, кадров {len(files)} → {out}")
        for f in files: print("  ", f)
        b.close()

if __name__ == '__main__':
    main()

File: tools/shots.js
#!/usr/bin/env node
/* Лист кадров для антислопометра.
   node shots.js <url> [--out DIR] [--w 1440] [--h 900] [--frames 8] [--wait 1200] [--full]
   Кадры снимаются со скроллом, анимации успевают отработать. */
const path = require('path'), fs = require('fs');
const NM = [
  path.join(process.env.HOME, '.claude/skills/antislop/tools/node_modules'),
  '/Users/alexeykolpikov/!-WORK/!-REFERENCES/Lebedev/tools/node_modules',
];
let chromium;
for (const p of NM) { try { ({ chromium } = require(path.join(p, 'playwright-core'))); break; } catch (e) {} }
if (!chromium) { console.error('playwright-core не найден'); process.exit(1); }

const a = process.argv.slice(2);
const url = a[0];
const opt = (k, d) => { const i = a.indexOf('--' + k); return i > -1 ? a[i + 1] : d; };
const flag = k => a.includes('--' + k);
const OUT = opt('out', path.join(process.env.TMPDIR || '/tmp', 'shots-' + new URL(url).hostname.replace(/\W/g, '-')));
const W = +opt('w', 1440), H = +opt('h', 900), FRAMES = +opt('frames', 8), WAIT = +opt('wait', 1200), SCALE = +opt('scale', 2);

(async () => {
  fs.mkdirSync(OUT, { recursive: true });
  const root = path.join(process.env.HOME, 'Library/Caches/ms-playwright');
  const cands = [];
  for (const d of fs.readdirSync(root).filter(d => d.startsWith('chromium')).sort().reverse()) {
    cands.push(
      path.join(root, d, 'chrome-mac-arm64/Chromium.app/Contents/MacOS/Chromium'),
      path.join(root, d, 'chrome-mac/Chromium.app/Contents/MacOS/Chromium'),
      path.join(root, d, 'chrome-headless-shell-mac-arm64/chrome-headless-shell'),
      path.join(root, d, 'chrome-headless-shell-mac/chrome-headless-shell'));
  }
  cands.push('/Applications/Google Chrome.app/Contents/MacOS/Google Chrome');
  const exe = cands.find(p => fs.existsSync(p));
  if (!exe) { console.error('браузер не найден'); process.exit(1); }
  const browser = await chromium.launch({
    headless: true, executablePath: exe,
    args: ['--hide-scrollbars', '--autoplay-policy=no-user-gesture-required', '--force-device-scale-factor=' + SCALE],
  });
  const page = await browser.newPage({ viewport: { width: W, height: H }, deviceScaleFactor: SCALE });
  await page.goto(url, { waitUntil: 'networkidle', timeout: 60000 }).catch(() => {});
  await page.waitForTimeout(2500);
  const total = await page.evaluate(() => document.body.scrollHeight);
  const step = Math.max(1, Math.round((total - H) / Math.max(1, FRAMES - 1)));
  const files = [];
  for (let i = 0; i < FRAMES; i++) {
    const y = i * step;
    await page.evaluate(v => window.scrollTo({ top: v, behavior: 'instant' }), y);
    await page.waitForTimeout(WAIT);
    const f = path.join(OUT, String(i + 1).padStart(2, '0') + '.png');
    await page.screenshot({ path: f });
    files.push(f);
  }
  if (!flag('nofull')) {
    const f = path.join(OUT, 'full.png');
    await page.screenshot({ path: f, fullPage: true });
    files.push(f);
  }
  await browser.close();
  console.log('высота страницы: ' + total + 'px, шаг ' + step + 'px');
  files.forEach(f => console.log(f));
})();
